python/ray/data/_internal/memory_tracing.py

```diff
@@ -0,0 +1,140 @@
+"""Utility for debugging object store memory eager deletion in Datasets.
+
+Enable with RAY_DATASET_TRACE_ALLOCATIONS=1.
+
+Basic usage is to call `trace_allocation` each time a new object is created, and call
+`trace_deallocation` when an object should be disposed of. When the workload is
+complete, call `leak_report` to view possibly leaked objects.
+
+Note that so called "leaked" objects will be reclaimed eventually by reference counting
+in Ray. This is just to debug the eager deletion protocol which is more efficient.
+"""
+
+from io import StringIO
+from typing import Dict
+
+import ray
+from ray.data.context import DatasetContext
+
+
+def trace_allocation(ref: ray.ObjectRef, loc: str) -> None:
```

Comment:
Could we create `loc` automatically for the caller with something akin to `inspect.getframeinfo(inspect.stack()[1])` or `inspect.getouterframes(inspect.currentframe())[2]`? Or we could reuse something let our `get_py_stack` utility: https://github.com/ray-project/ray/blob/2bedf2490d7b11c2728c3fc8c57cf337a5e296ef/python/ray/_raylet.pyx#L1375-L1415

Reply:
Probably. I'll add this as a TODO, as it isn't really a pain point.